File: app/src/main/java/com/example/roombooking/activities/BaseActivity.java
package com.example.roombooking.activities;

import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;

import com.example.roombooking.R;
import com.example.roombooking.controllers.SwitchFragmentController;
import com.example.roombooking.fragments.BaseFragment;

public abstract class BaseActivity extends AppCompatActivity
{
	/**
	 * switch to a fragment, and add it to back stack
	 *
	 * @param pBaseChildFragment: the instance of fragment that want to show
	 */
	protected <T extends BaseFragment> void switchFragment (T pBaseChildFragment)
	{
		SwitchFragmentController.switchFragment(R.id.containerView, getSupportFragmentManager(), pBaseChildFragment, null, true);
	}

	/**
	 * switch to a fragment, and add it to back stack
	 *
	 * @param pBaseChildFragment: the instance of fragment that want to show
	 */
	protected <T extends BaseFragment> void switchFragment (T pBaseChildFragment, Bundle dataBundle)
	{
		SwitchFragmentController.switchFragment(R.id.containerView, getSupportFragmentManager(), pBaseChildFragment, dataBundle, true);
	}

	/**
	 * switch to a fragment, and add it to back stack if defined
	 *
	 * @param pBaseChildFragment: the instance of fragment that want to show
	 * @param isAddToBackStack: add pBaseChildFragment to back stack if needed
	 */
	protected <T extends BaseFragment> void switchFragment (T pBaseChildFragment, Bundle dataBundle, boolean isAddToBackStack)
	{
		SwitchFragmentController.switchFragment(R.id.containerView, getSupportFragmentManager(), pBaseChildFragment, dataBundle, isAddToBackStack);
	}

	@Override
	public void onBackPressed()
	{
		BaseFragment mFragment = SwitchFragmentController.getCurrentFragment(getSupportFragmentManager());
		if(mFragment != null)
			mFragment.onBackPressed();

		if(getSupportFragmentManager().getBackStackEntryCount() == 0)
			finish();
	}
}

File: app/src/main/java/com/example/roombooking/activities/BaseActivity.kt
package com.example.roombooking.activities

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import com.example.roombooking.R
import com.example.roombooking.controllers.SwitchFragmentController.getCurrentFragment
import com.example.roombooking.controllers.SwitchFragmentController.switchFragment
import com.example.roombooking.fragments.BaseFragment

abstract class BaseActivity : AppCompatActivity() {
    /**
     * switch to a fragment, and add it to back stack
     *
     * @param pBaseChildFragment: the instance of fragment that want to show
     */
    protected fun <T : BaseFragment?> switchFragment(pBaseChildFragment: T) {
        switchFragment(R.id.containerView, supportFragmentManager, pBaseChildFragment, null, true)
    }

    /**
     * switch to a fragment, and add it to back stack
     *
     * @param pBaseChildFragment: the instance of fragment that want to show
     */
    protected fun <T : BaseFragment?> switchFragment(pBaseChildFragment: T, dataBundle: Bundle?) {
        switchFragment(R.id.containerView, supportFragmentManager, pBaseChildFragment, dataBundle, true)
    }

    /**
     * switch to a fragment, and add it to back stack if defined
     *
     * @param pBaseChildFragment: the instance of fragment that want to show
     * @param isAddToBackStack: add pBaseChildFragment to back stack if needed
     */
    protected fun <T : BaseFragment?> switchFragment(pBaseChildFragment: T, dataBundle: Bundle?, isAddToBackStack: Boolean) {
        switchFragment(R.id.containerView, supportFragmentManager, pBaseChildFragment, dataBundle, isAddToBackStack)
    }

    override fun onBackPressed() {
        val mFragment = getCurrentFragment(supportFragmentManager)
        mFragment?.onBackPressed()
        if (supportFragmentManager.backStackEntryCount == 0) finish()
    }
}